### ISSUE TITLE
Pushy should provide a way of getting a notification that couldn't be queued

### ISSUE THREAD

**Issue by philipwhiuk** [relayrides/pushy#439]
The documentation suggests the best practice is:

```java

pushNotificationResponseHandler = 
   new FutureListener<PushNotificationResponse<SimpleApnsPushNotification>>() {
      public void operationComplete() {
           try {
                responseFuture.get()
           } catch (ClientNotConnectedException e) {
                  apnsClient.getReconnectionFuture.await()
                  //Resend!?
           }
      }
   };

future = apnsClient.sendNotification(pushNotification);
future.addListener(pushNotificationResponseHandler);
```

But there's no way to get a reference in a `FutureListener<PushNotificationResponse<SimpleApnsPushNotification>>` to the push notification which has failed to send.

Handling this case involves one of:

1. Making assumptions about Pushy (i.e. is the queue processed in order - can I just maintain my own queue of unsent notifications and just assume the failed one is the head of queue?
2. Create a handler per notification.

Of these, 1 looks decidedly dodgy and 2 is unscalable.

What's the recommended way of resending `PushNotification`s that fail due to an `ClientNotConnectedException`.

It would be nice if either `ApnsClient` or the `ClientNotConnectedException` itself exposed the `PushNotification`.

**Comment by jchambers** [comment]
> …can I just maintain my own queue of unsent notifications and just assume the failed one is the head of queue?

We make no guarantees about order, so the queue approach won't work.

> Create a handler per notification.

This has generally worked well so far, even for heavy use cases. I agree that there's an opportunity to decrease overall memory usage by adding a mechanism to get the original push notification when send futures fail (thereby paving the way for reuse of a single listener instance), but don't think I would describe the current situation as "unscalable;" either way, we're still talking about _O(n)_ memory usage, where _n_ is the number of notifications "in flight."

> What's the recommended way of resending PushNotifications that fail due to an ClientNotConnectedException.

The most common approach is to add the listener as an anonymous implementation that refers to the sent notification (please excuse the lazy syntax here):

```java
final Future<> sendFuture = client.sendNotification(exampleNotification);
sendFuture.addListener(new FutureListener<PushNotificationResponse<SimpleApnsPushNotification>>() {
      public void operationComplete() {
           try {
                responseFuture.get()
           } catch (ClientNotConnectedException e) {
                  apnsClient.getReconnectionFuture.await()

                  // exampleNotification is in scope
                  client.sendNotification(exampleNotification);
           }
      }
);
```

Alternatively, you could use a named listener that takes a push notification as a constructor:

```java
final Future<> sendFuture = client.sendNotification(exampleNotification);
sendFuture.addListener(new ListenerThatKnowsAboutASpecificNotification(exampleNotification));
```

In short, this seems like a good thing to think about, but I think other things are likely to take priority.

**Comment by jchambers** [comment]
We have a proposal for a potential solution in #535. What do you think?
